shap/plots/_scatter.py

```diff
@@ -528,7 +426,106 @@ def scatter(
     if show:
         with warnings.catch_warnings():  # ignore expected matplotlib warnings
             warnings.simplefilter("ignore", RuntimeWarning)
-            pl.show()
+            plt.show()
+    else:
+        return ax
+
+
+def _parse_limit(ax_limit: LimitSpec, ax_values: np.ndarray, is_shap_axis: bool) -> float | None:
+    """Handle axis limits in "percentile(float)" format or from Explanation objects"""
+    if isinstance(ax_limit, str):
+        if not ax_limit.startswith("percentile(") or not ax_limit.endswith(")"):
+            raise ValueError("Only strings of the format `percentile(x)` are supported.")
+        return np.nanpercentile(ax_values, float(ax_limit[11:-1]))
+    if isinstance(ax_limit, Explanation):
+        # Expect Explanation aggregated to a single value, e.g. `explanation[:, "feature_name"].percentile(20)`
+        # Extract relevant attribute, depending if x- or y-axis
+        return float(ax_limit.values) if is_shap_axis else float(ax_limit.data)
+    # Else, should be float or None
+    return ax_limit
+
+
+def _suggest_buffered_limits(ax_min: float | None, ax_max: float | None, values: np.ndarray) -> tuple[float, float]:
+    """If either limit is None, suggest suitable value including a buffer either side"""
+    nan_max = np.nanmax(values) if ax_max is None else ax_max
+    nan_min = np.nanmin(values) if ax_min is None else ax_min
+    buffer = (nan_max - nan_min) / 20
+    if ax_min is None:
+        ax_min = float(nan_min - buffer)
+    if ax_max is None:
+        ax_max = float(nan_max + buffer)
+    return ax_min, ax_max
+
+
+def _suggest_x_jitter(values: np.ndarray) -> float:
+    """Suggest a suitable x_jitter value based on the unique values in the feature"""
+    unique_vals = np.sort(np.unique(values))
+    try:
+        # Identify the smallest difference between uniqu evalues
+        diffs = np.diff(unique_vals)
+        min_dist = np.min(diffs[diffs > 1e-8])
```

Comment:
cmiiw, but isn't it possible that `diffs` here will all contain `diffs > 1`?
In which case, your implementation would be different from the version before refactoring, which essentially does a `min(*diffs, 1)` 

![image](https://github.com/user-attachments/assets/f44eda91-5f8b-4896-b1cd-7ccf9f915d8e)

Reply:
Yes, I believe that is correct. However, I think this change is desirable.

Up until 4 days ago, the `min_dist` was simply the minimum distance between points:

```
min_dist = np.inf
for i in range(1, len(vals)):
    d = vals[i] - vals[i - 1]
    if d > 1e-8 and d < min_dist:
        min_dist = d
```

However, this was updated recently by #3706 to fix the plotting of categorical variables. The `min_dist=1` default is fairly arbitrary, intended for the case where there are no numeric values to compare. When reviewing that PR I missed that it would actually always limit `min_dist` to 1, so here I propose reverting to the prior behaviour :)

Reply:
got it, thanks for the explanation!